lib/lookup-table/include/LookupTable.h

```diff
@@ -0,0 +1,68 @@
+#pragma once
+#include <iterator>
+#include <optional>
+#include <unordered_map>
+
+// This a LUT type. It is not meant to be mutated after initilization.
+template <class Key, class Value, class Hash = std::hash<Key>, class KeyEqual = std::equal_to<Key>,
+          class Allocator = std::allocator<std::pair<const Key, Value>>>
+class LookupTable {
+ private:
+  const std::unordered_map<Key, Value, Hash, KeyEqual, Allocator> _unordered_map;
+
+ public:
+  using key_type        = Key;
+  using mapped_type     = Value;
+  using value_type      = std::pair<const Key, Value>;
+  using size_type       = size_t;
+  using difference_type = std::ptrdiff_t;
+  using hasher          = Hash;
+  using key_equal       = KeyEqual;
+  using allocator_type  = Allocator;
+  using reference       = value_type &;
+  using const_reference = const value_type &;
+  using pointer         = std::allocator_traits<Allocator>::pointer;
+  using const_pointer   = std::allocator_traits<Allocator>::const_pointer;
+
+  LookupTable() = delete;
+
+  // range based constructor
+  template <typename InputIterator>
+  LookupTable(InputIterator first, InputIterator last) : _unordered_map(first, last) {}
+  // copy constructor and assignment
+  LookupTable(const LookupTable &copy) = default;
+  LookupTable &operator=(const LookupTable &) = default;
+  // move constructor and assignment
+  LookupTable(LookupTable &&move) = default;
+  LookupTable &operator=(LookupTable &&) = default;
+  // initializer list constructor
+  LookupTable(std::initializer_list<value_type> init, size_type bucket_count = 0, const Hash &hash = Hash(),
+              const key_equal &equal = key_equal(), const Allocator &alloc = Allocator())
+      : _unordered_map(init, bucket_count, hash, equal, alloc) {}
+
+  typename std::unordered_map<Key, Value>::iterator begin() {
+    return _unordered_map.begin();
+  }
+  typename std::unordered_map<Key, Value>::iterator end() {
+    return _unordered_map.end();
+  }
+
+  std::size_t size() const {
+    return _unordered_map.size();
+  }
+  std::size_t max_size() const {
+    return _unordered_map.max_size();
+  }
+
+  bool empty() const {
+    return _unordered_map.empty();
+  }
+
+  std::optional<Value> at(const Key &key) const {
+    if (_unordered_map.find(key) == _unordered_map.end()) return std::nullopt;
+    return _unordered_map.at(key);
```

Comment:
```suggestion
    return _unordered_map.at(key)?_unordered_map.find(key) !=_unordered_map.end():std::nullopt;
```

oneliner

Reply:
another possibility thats maybe more readable

```suggestion
    bool value_exists = _unordered_map.find(key) == _unordered_map.end();
    return _unordered_map.at(key)?value_exists:std::nullopt;
```

Reply:
i think u meant return value_exists ? map.at(key) : nullopt. Even then, this wont work bc ternary operator must have same time on either side of colon

Reply:
Ye. Typo in the order.

Wdym same time?

Reply:
oops same type

Reply:
U just have to wrap the true value in a std::optional

Reply:
👍 :O